src/tr1/game/ui/widgets/paginator.c

```diff
@@ -0,0 +1,228 @@
+        .get_width = (UI_WIDGET_GET_WIDTH)M_GetWidth,
+        .get_height = (UI_WIDGET_GET_HEIGHT)M_GetHeight,
+        .set_position = (UI_WIDGET_SET_POSITION)M_SetPosition,
+        .free = (UI_WIDGET_FREE)M_Free,
+    };
+
+    self->outer_stack = UI_Stack_Create(
+        UI_STACK_LAYOUT_VERTICAL, UI_STACK_AUTO_SIZE, UI_STACK_AUTO_SIZE);
+
+    const char *upper_title = String_ToUpper(title);
+    self->title =
+        UI_Label_Create(upper_title, UI_LABEL_AUTO_SIZE, TEXT_HEIGHT_FIXED);
+    Memory_FreePointer(&upper_title);
+    UI_Stack_AddChild(self->outer_stack, self->title);
+
+    self->top_spacer = UI_Spacer_Create(TITLE_MARGIN, TITLE_MARGIN);
+    UI_Stack_AddChild(self->outer_stack, self->top_spacer);
+
+    const char *wrapped =
+        String_WordWrap(text, Text_GetMaxLineLength() - PADDING_SCALED);
+    self->page_content = String_Paginate(wrapped, max_lines);
+    self->current_page = 0;
+    Memory_FreePointer(&wrapped);
+
+    self->text = UI_Label_Create(
+        *(char **)Vector_Get(self->page_content, 0), UI_LABEL_AUTO_SIZE,
+        UI_LABEL_AUTO_SIZE);
+    UI_Stack_AddChild(self->outer_stack, self->text);
+
+    if (self->page_content->count > 1) {
+        self->bottom_spacer = UI_Spacer_Create(TITLE_MARGIN, TITLE_MARGIN * 3);
+        UI_Stack_AddChild(self->outer_stack, self->bottom_spacer);
+
+        self->bottom_stack = UI_Stack_Create(
+            UI_STACK_LAYOUT_HORIZONTAL, UI_STACK_AUTO_SIZE, UI_STACK_AUTO_SIZE);
+        UI_Stack_AddChild(self->outer_stack, self->bottom_stack);
+
+        self->left_arrow =
+            UI_Label_Create("\\{button left}", 22, TEXT_HEIGHT_FIXED);
+        self->right_arrow =
+            UI_Label_Create("\\{button right}", 16, TEXT_HEIGHT_FIXED);
+        self->right_arrow_spacer = UI_Spacer_Create(6, TEXT_HEIGHT_FIXED);
+        self->page_label =
+            UI_Label_Create("", UI_LABEL_AUTO_SIZE, UI_LABEL_AUTO_SIZE);
+        UI_Stack_AddChild(self->bottom_stack, self->left_arrow);
+        UI_Stack_AddChild(self->bottom_stack, self->page_label);
+        UI_Stack_AddChild(self->bottom_stack, self->right_arrow_spacer);
+        UI_Stack_AddChild(self->bottom_stack, self->right_arrow);
+
+        UI_Stack_SetHAlign(self->bottom_stack, UI_STACK_H_ALIGN_RIGHT);
+    }
+
+    self->window = UI_Window_Create(
+        self->outer_stack, DIALOG_PADDING, DIALOG_PADDING, DIALOG_PADDING * 2,
+        DIALOG_PADDING);
+
+    // Ensure minimum width for page navigation as text content may be empty.
+    int32_t max_width =
+        MAX(self->page_content->count == 1 ? 0 : 100,
+            UI_Label_MeasureTextWidth(self->title));
+    int32_t max_nav_width = 0;
+    int32_t max_height = 0;
+    for (int32_t i = 0; i < self->page_content->count; i++) {
+        M_SelectPage(self, i);
+        max_width = MAX(max_width, UI_Label_MeasureTextWidth(self->text));
+        max_height = MAX(max_height, UI_Label_MeasureTextHeight(self->text));
+        if (self->bottom_stack != NULL) {
```

Comment:
I think in this case, `UI_Label_MeasureTextWidth(self->page_label)` will default to 0, no?

Reply:
This handles (silly) cases where you have a very short title, very short page entries, but many pages, such that the biggest element is the navigation itself. So we cycle through and may find `<- 49 / 50 ->` is the widest label. The page nav is hidden if there is only one page, and in that case bottom_stack is null.

I'm open to tidying this up though if you've any thoughts on an improved implementation.

Reply:
No it's just that if the underlying page nav label is an empty string it should still be 0. If it's filled with the pager, I'm fine with the current code.